### PR TITLE
feat(xtask): 'changes' xtask

### PR DESCRIPTION
**Pull request by epage**
This will report what commits affect each publishable package and serves as an alternative to rust-lang/cargo#12085.

CI can run this as `cargo changes HEAD~ HEAD^2` and that will capture all of the commits within the PR (this is the range I used in `committed`).

There is still work needed to integrate this with an action; this is just a rough base-line implementation.

I originally tried to use `cargo package --list` to determine what files belong to what packages but that was taking too long for some reason (it works well in `cargo release` for my crates).  We can either look into that in the future or do our own heuristics to filter out content.

### CODE REVIEW

**Review by weihanglo — COMMENTED** [review]
Thanks for the fast prototyping. This looks pretty complete to me!

Some thoughts:

* The cargo project doesn't adopt git conventional commit message pattern. Not sure if it is worthy or we should enforce that.
* Perhaps switch to `gix` if we plan to drop `git2` someday.
* If we merge this or else, should we remove `xtask-unpublished`?
* The code is indeed in a high quality. I can't be pickier at code level. However, total lines of code usually correlate to maintenance cost. I am a bit worried it's a bit overkill when comparing to what can be done in shell scripts.
* This is a good demonstration of how people can invent something interesting with `cargo` and other libraries. I love the pretty status output!

**Comment by weihanglo** [comment]
> At the moment, this is also written in a way that I could see it being useful for writing up changelogs, so we might be able to get more mileage out of it.

This made me wonder we would like to change each PR title before `r+`. It will make more sense for conventional commit messages.

> Unsure, in part because these are experiments but in part because I don't fully remember what all building blocks I originally intended...

My fault. I should ask your original intent first for `xtask-unpublished` but what I did was just guessing 😅.

To be clear, I see the values in these conventions and the command. However, it feels like going too far away from #12033. I was hoping there could be a way to catch not-yet-bumped versions in CI ASAP, so that we can start splitting internal building blocks into subcrates.

**Comment by ehuss** [comment]
r? weihanglo

I think Weihang is following this more closely than I am. I haven't really reviewed this in detail, but the general concept seems good.

File: crates/xtask-changes/src/xtask.rs [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+#[derive(Clone, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub struct CommitDescription {
+    id: git2::Oid,
+    short_id: String,
+    summary: String,
+    message: String,
+    paths: std::collections::BTreeSet<std::path::PathBuf>,
+}
+
+impl CommitDescription {
```
Comment (weihanglo) [inline crates/xtask-changes/src/xtask.rs:305]: I would probably suggest to remove conventional commit part. Let's focus on the original goal for now.

**Review by weihanglo — COMMENTED** [review]
This looks fine for listing changes. However, I still don't know how it works with requiring user to do a version bump. Did it really give user the information that they need bumping a version for a crate?

**Comment by weihanglo** [comment]
**[Not a blocker]** I still lean toward a similar approach like #12126 as a immediate guard rail. We can then properly design and expand this automation, thinking more on what we want for contributors and CI.

**Comment by epage** [comment]
Based on the direction of #12395, I think we can close this.  What information we can learn from here has been called out in that other PR.

**Pull request closed without merging**